Update operation mode during init

aronsky opened the pull request:
## Description:

Upon restart, Home Assistant doesn't turn the thermostat on, even if there's a default temperature set in the configuration, and based on the current temperature, the thermostat should be on. This PR fixes the behavior by calling the appropriate method during the thermostat's initialization.

## Checklist:

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

aronsky committed:
Update operation mode during init

pvizeli commented:
I don't know is that needed, but you need use the EVENT_HOMEASSISTANT_START for that.

aronsky commented:
It's not critical, but it's certainly a bug - when I start Home Assistant and open the frontend, it shows me the thermostat (with its set temperature and the current room temperature), and it's at idle.

To get it working, I have to change the set temperature. Alternatively, I can wait for the temperature to change a bit (at which point the sensor changed event will fire, and that'll update the operation mode). Like I said - not critical, but certainly not the correct operation.

I didn't bother with ```EVENT_HOMEASSISTANT_START```, since there's already initialization code that runs under the correct conditions (when the sensor's state is available, see the line above my change). Would you rather I move that code into a subfunction which will be called when ```EVENT_HOMEASSISTANT_START``` fires (like here: https://github.com/home-assistant/home-assistant/blob/9cad9c19f8a29c0173cb7672fd3ad95d91c34ee1/homeassistant/components/keyboard_remote.py#L67)? It's possible, but a bit of an overhead, I think - after all, this code is already inside the constructor, isn't that an appropriate place for it?

balloob commented on homeassistant/components/climate/generic_thermostat.py:
```diff
@@ -107,6 +107,7 @@ def __init__(self, hass, name, heater_entity_id, sensor_entity_id,
         sensor_state = hass.states.get(sensor_entity_id)
         if sensor_state:
             self._async_update_temp(sensor_state)
+            self._async_control_heating()
```
Please check that the switch has already been set up.

pvizeli replied:
Make not more sense to move this into EVENT_HOMEASSISTANT_START?

aronsky replied:
Does ```EVENT_HOMEASSISTANT_START``` fire after all the switches and sensors have been set up?

balloob replied:
Yes

balloob replied:
And since this platform has a dependency on switch and platform, everything that is not discovered should also have been set up already.

aronsky replied:
Done.

aronsky committed:
Use EVENT_HOMEASSISTANT_START.

Moved the call to _async_control_heating out of the constructor and
into the platform setup code, in order to make correct use of the
EVENT_HOMEASSISTANT_START event and to be sure that the switches/sensors
have already been set up.

aronsky committed:
Moved the listener to EVENT_HOMEASSISTANT_START.

Moved the listener to EVENT_ASSISTANT_START into the constructor of the
thermostat component, in order to avoid calling a private method from
outside the class.

pvizeli commented on homeassistant/components/climate/generic_thermostat.py:
```diff
@@ -108,6 +109,14 @@ def __init__(self, hass, name, heater_entity_id, sensor_entity_id,
+        def _start_generic_thermostat(_event):
+            self._async_control_heating()
+
+        hass.bus.async_listen_once(
```
move this into coro `async_added_to_hass` and add @callback to your function

aronsky replied:
Done.

aronsky committed:
Moved event registration into

pvizeli commented:
For me is okay. @balloob ?

balloob reviewed (changes requested):
Please add a test to verify the new behavior

andrey-git commented:
@aronsky Lev, do you plan to continue working on this PR?

aronsky commented:
@andrey-git Andrey, yeah, I've just been busy with work in the last couple of weeks, and haven't had a chance to get to writing the tests.

pvizeli commented:
Please reopen it after you add tests.
